### DISCUSSION TITLE
Extension methods in namespaces

### DISCUSSION THREAD

**Discussion by csaba-sagi-sonarsource** [dotnet/csharplang#4942] [Language Ideas]
# Summary

Allow extension methods directly in namespaces.

The semantics are that if such extension methods are found, than a static class would be generated in that namespace that would contain the extension method. This would also mean that instead of having N different static classes in a namespace that contain the extension methods, all the extension methods from the same namespace could be part of a single static class. 

# Motivation

There is the boilerplate code (the static class) that contains the extension method, which is not needed, in the sense that the  method is not intended to be invoked like: `StaticClass.ExtensionMethod(arg1, arg2)`, so the class name can be anything. The primary goal of the feature therefore is to allow C# programs without the unnecessary boilerplate around them, and let the static class be generated instead.

# Current way to write an extension method

``` c#
namespace ExtensionMethods
{
    public static class MyExtensions
    {
        public static int WordCount(this string str) =>
            str.Split(new char[] { ' ', '.', '?' }, StringSplitOptions.RemoveEmptyEntries).Length;
    }
}
```
# Proposed way to write an extension method

``` c#
namespace ExtensionMethods;

public static int WordCount(this string str) =>
    str.Split(new char[] { ' ', '.', '?' }, StringSplitOptions.RemoveEmptyEntries).Length;
```

**Comment by svick** [comment]
I believe eliminating boilerplate code is important if it has high impact: for example if it's something that occurs very often (expression-bodied members), if it can eliminate a lot of code (records) or if it's literally the first thing seen by someone learning C# (top-level programs).

But none of that applies here: the amount of code eliminated is small (3 LOC) and it doesn't happen very often in most codebases. So I don't see why doing this on its own would be worthwhile. It could make sense as part of more general top-level functions (https://github.com/dotnet/csharplang/issues/3117), but that proposal has been rejected.

**Reply by FaustVX** [reply]
And also, you won't be able to use `using static`.
